### ISSUE TITLE
Improve event insert/worker performance

### ISSUE THREAD

**Issue by timgl** [PostHog/posthog#809]
**Describe the bug**
We have a user that's doing 40 req/sec. The workers can't keep up with the volume of events, so the Redis cache fills up and then falls over

At the moment their setup is
1. 3 web dynos 2X
1. 15 worker dynos 2X
1. 1GB Redis cache
1. Standard-1 postgres database

As you can see [in this test](https://github.com/PostHog/posthog/blob/master/posthog/tasks/test/test_process_event.py#L19), we do 19 queries for each event insert (with autocapture).

**ideas**
- I think the query that [finds the relevant actions for an event](https://github.com/PostHog/posthog/blob/master/posthog/models/event.py#L239) is slow, so it might be worth using the same pattern we have for Cohorts to calculate them after the fact.
- For this specific user, the dom tree is probably unique for a lot of events so have to be re-inserted. We already bulk-insert all of the Elements, but there might still be gains to be had here

**Comment by timgl** [comment]
After having this user remove all actions on the instance, it's easily processing 40req/sec with only 3 worker dynos and redis usages is down to 5mb (so no events are backing up).

This suggests we need to use the same pattern as in cohorts to precalculate actions (or find another solution) instead of doing it on each event save

**Comment by fuziontech** [comment]
So far here are my findings so far:

**TL;DR => An index on properties is a quick win here decreasing query time by ~20%.**

For the tests I used the following Actions configs:
![image](https://user-images.githubusercontent.com/391319/82632495-3c356080-9bad-11ea-940c-fd8f13a3c8ee.png)
NOTE: pageviews action had 3k views at time of testing, not 300

- Events are about 25-50% faster than actions with 5 queries per insert for my config. The range of latencies from queries was ~15ms:
![image](https://user-images.githubusercontent.com/391319/82629771-8a466600-9ba5-11ea-8186-879ba576f12d.png)
Wall clock time for running just the task portion of the celery worker was ~80ms

- Actions (like $pageview) are 25-50% slower than events with 8 queries per insert for my config. The range of latencies from queries was 25-30ms:
![image](https://user-images.githubusercontent.com/391319/82629786-9cc09f80-9ba5-11ea-98a5-39dd59fa5fbb.png)
The interesting thing here is wall clock time for the celery task was ~180ms

- Partitioning does not affect the performance of these queries locally.

- Doubling the number of Actions with fuzzy matching URLs has no material impact on total query time
![image](https://user-images.githubusercontent.com/391319/82630939-10b07700-9ba9-11ea-80a5-8eb50ec889fd.png)
Resulting sample latency - Still 25-30ms:
![image](https://user-images.githubusercontent.com/391319/82630927-07270f00-9ba9-11ea-813c-88699473494f.png)

- Removing all `URL Match` actions significantly increases speed of query as we've seen with customers in production bringing the speed:
![image](https://user-images.githubusercontent.com/391319/82631124-9d5b3500-9ba9-11ea-9832-6085cf0955a8.png)
Result:
![image](https://user-images.githubusercontent.com/391319/82631212-d1cef100-9ba9-11ea-83fe-8ee63111d37b.png)
This one makes sense since the query that is built hits the `properties` jsonb field not indexed with a `LIKE`.

- The best improvement of database time has been an index on `properties`. I was worried that this would be a net-negative because of insert cost but overall there's a ~5-10ms improvement in DB time. Which is still ~20% improvement
![image](https://user-images.githubusercontent.com/391319/82632523-4fe0c700-9bad-11ea-8b8e-6abcaffa85c6.png)


**TODO** - Things that are unexplored and could be opportunities:
- Research & talk with Tim on how we are cohorting users as a potential solution
- Look into ways we can reduce the overhead of the task itself. As it is right now 20% of the time is spent in the DB, the rest is in the celery task. This is where I think we could have the largest impact. I am very curious as to where the 80% of the time is going.
- If it comes to it, we did something like this at Disqus: https://github.com/gocelery/gocelery reimplementing the event consumer in go. This will probably be something we will need to do at some point, but we have to have a hard think about how we would want keep models in sync. I'd like to avoid this until we have a chance to chat about long term architecture.
